####-Pulse-Channels-IR-and-Compiler.md

```diff
@@ -0,0 +1,145 @@
+The main goal is to allow for clear and simple backend agnostic (but not architecture agnostic) pulse templates.
+To do this, one must move away from the `Channel`s setup, which can't be used without specific backend mapping.
+
+The new Pulse IR and compiler will support the transition, and provide a unified compilation tool for every need and desired output format.
+
+No API breaking changes are needed. New and legacy options can coexist.
+
+## Detailed Design
+### Channel rework
+To replace the legacy Channels we propose to specify instructions in terms of `LogicalElement`s and `Frame`s:
+
+- `LogicalElement` - every type of element in the HW which the user can control ("play pulses on"). For example, a qubit is a `LogicalElement`, and not the specific port used to drive it. The most notable example of a logical element is of course the `Qubit`, but in the future one can imagine dedicated couplers, flux controls and so on.
+- `Frame` - a combination of frequency and phase. Subclasses used to identify the frame with a backend default. Notable examples are `QubitFrame` and `MeasurementFrame` associated with the default driving and measurement frequencies (respectively) of a qubit.
+Using these objects, the above code takes the form:
+
+```
+with builder.build() as ecr_schedule_q3q4:
+    with align_sequential():
+        with align_left():
+            Play(GaussianSquare(...), Qubit(3),QubitFrame(4))
+            Play(GaussianSquare(...), Qubit(3),QubitFrame(3))
+        Play(Gaussian(...), Qubit(3),QubitFrame(3))
+        with align_left():
+            Play(GaussianSquare(...), Qubit(3),QubitFrame(4))
+            Play(GaussianSquare(...), Qubit(3),QubitFrame(3))
+```
+
+The new code not only allows to work with logical backend-agnostic qubits, but is also much clearer in conveying the actual actions in play (acting on qubit 3 with the frame of qubit 4 vs with the frame of qubit 3).
+
+On top of the `LogicalElement` and the `Frame`, another layer of `MixedFrame` will be introduced.
+The `MixedFrame` is simply a combination of a LogicalElement and a Frame, and in many ways is similar to the legacy `Channel`.
+The difference being that a `MixedFrame` does not depend on the backend mapping and has clear relations with other `MixedFrames` as opposed to a `Channel` which hasn't.
+
+It should be noted that custom `Frame` and `MixedFrame` will make it easier to control qudit experiments.
+
+### Pulse IR
+The Pulse IR will provide the framework for the compiler to work on and perform the necessary compilation steps. The main components of the IR will be:
+
+- IR Block - A block of instructions (or nested blocks) with an alignment context. This structure is similar to the structure of ScheduleBlock and needed to support one of the main tasks of the compiler - scheduling.
+- IR Instruction - A pulse program instruction. Typically, it will be initialized without concrete timing, and will be scheduled during the compiler operation.
```

Comment:
The instructions will be grouped in the IR objects, so I think that will make a composite pattern a bit redundant here?

Reply:
Could be. A good point of using composite pattern is that we can remove branching or dispaching which could make code more readable (originally we have implemented `Schedule` and `Instruction` by using this pattern). Maybe this doesn't fit in here because `PulseIR` doesn't need time point.

Reply:
It seems to me like everything we can do with a composite pattern we can also do by acting on `PulseIR` objects? We can revisit this later to see if we can improve code readability.